### PR TITLE
Fix dnsmasq/IPSET snapshot recovery, initialize IPSET_START_STOPPED, and tighten related tests

### PR DESCRIPTION
**Pull request by jumpsmm7**
### Motivation

- Prevent a stale restore-only snapshot (created when dnsmasq backup copy failed) from being treated as a real transaction and restoring older IPSET/YAML over newer live state.
- Avoid BusyBox `[` bad-number failures when `IPSet_Disable_Managed_For_Start_Locked()` examines `IPSET_START_STOPPED` while it may be unset.
- Strengthen and extend focused regression tests to cover the backup-copy failure, cleanup-only snapshots, service-stop failures, iptables-query contracts, helper extraction safety, and other edge cases discovered during review.

### Description

- Treat snapshots whose directory contains `restore.pending` but neither `config.pending` nor `config.restored` as cleanup-only in `dnsmasq_ipset_state_recover_pending()` and skip calling `dnsmasq_ipset_state_restore` for those snapshots while preserving retry/marker behavior on cleanup failure.
- Initialize `IPSET_START_STOPPED` defensively (`IPSET_START_STOPPED="${IPSET_START_STOPPED:-0}"`) at the start of `IPSet_Disable_Managed_For_Start_Locked()` to ensure numeric tests never see an unset variable.
- Add and adjust focused regression tests and fixtures: create `config.restored` associations in dnsmasq snapshots where appropriate, add a cleanup-only snapshot scenario for failed backup-copy cleanup, ensure cleanup-only recovery does not overwrite newer IPSET/YAML, and add a stopped-service refresh check that verifies no numeric-test error or unintended service stop.
- Improve test harnesses: make `run_uninstall_test()` accept a configurable `stop` status and update `agh_stop()` to return it, change the `iptables()` stub to record the invoked query and add `assert_iptables_query()` to validate the exact command from the parent shell, guard `event_scripts_snapshot` extraction with an explicit end-boundary check, replace `$$`-derived temp paths with `mktemp -d` for exclusive temporary directories, add persisted `ap` and `bridge` mode cases, add nested-flock test coverage, and silence an unverified `rollback_result_notice()` mock in one test.

### Testing

- Ran POSIX syntax checks: `sh -n AdGuardHome.sh` and `sh -n` for all modified test scripts and they returned no syntax errors.
- Executed focused regression suites and end-to-end test scripts under the repo test harness with timeouts; the following ran and passed: `tests/dnsmasq-lan-mode.sh`, `tests/adguardhome-proc-settings.sh`, `tests/adguardhome-runtime-mode-helpers.sh`, `tests/dns-startup-handoff.sh`, `tests/installer-dns-environment-failure.sh`, `tests/installer-event-script-modes.sh`, `tests/installer-ipset-allowed-mode-fallback.sh`, `tests/installer-ipset-setup-save-failure.sh`, `tests/installer-upgrade-runtime-defaults.sh`, `tests/ipset-lan-mode.sh`, `tests/ipset-lock-security.sh`, and `tests/ipset-lan-mode.sh` (full run under timeout).
- Confirmed the focused scenarios that motivated the change: failed dnsmasq backup-copy cleanup leaves a restore-only marker that is treated cleanup-only and does not restore stale IPSET/YAML, and the stop-failure uninstall path preserves the installation and restores the domain; all related assertions passed in the test runs.
- ShellCheck was not available in the validation environment, so `shellcheck -s sh` was not executed here.

------
[Codex Task](https://chatgpt.com/codex/cloud/tasks/task_e_6a9922b8d5f08329bdc4a48a22d49978)

### CODE REVIEW

**Comment by jumpsmm7** [comment]
@coderabbitai approve

**Comment by jumpsmm7** [comment]
@qodo-ai stop

**Comment by jumpsmm7** [comment]
@coderabbitai pause

File: AdGuardHome.sh [hunk 1]
```diff
@@ -1047,6 +1047,14 @@ dnsmasq_ipset_state_recover_pending() {
 			return 1
 		fi
 		[ -f "${SNAPSHOT_DIR}/restore.pending" ] && [ ! -L "${SNAPSHOT_DIR}/restore.pending" ] || return 1
+		if [ ! -e "${SNAPSHOT_DIR}/config.pending" ] && [ ! -L "${SNAPSHOT_DIR}/config.pending" ] &&
+			[ ! -e "${SNAPSHOT_DIR}/config.restored" ] && [ ! -L "${SNAPSHOT_DIR}/config.restored" ]; then
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply AdGuardHome.sh]: @coderabbitai explain this issue as a prompt for an AI Agent

File: tests/adguardhome-proc-settings.sh [hunk 2]
```diff
@@ -389,6 +393,10 @@ run_uninstall_test restart-failure 1 1 && fail 'restore and restart failures did
 [ -d "${TMP_ROOT}/uninstall-restart-failure" ] || fail 'restart failure removed retained installation path'
 [ "$(sed -n '3p' "${TMP_ROOT}/events-restart-failure")" = 'ERROR Unable to restore installer-managed kernel settings.' ] || fail 'restart failure obscured restoration error'
 [ "$(sed -n '4p' "${TMP_ROOT}/events-restart-failure")" = start ] || fail 'restart failure was not exercised'
+run_uninstall_test stop-failure 0 0 usable 0 1 1 0 1 && fail 'service stop failure did not abort uninstall'
+[ -d "${TMP_ROOT}/uninstall-stop-failure" ] || fail 'service stop failure removed the installation path'
+! grep -qx remove "${TMP_ROOT}/events-stop-failure" || fail 'service stop failure removed installation files'
+grep -qx domain-restore "${TMP_ROOT}/events-stop-failure" || fail 'service stop failure did not restore the LAN domain'
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/adguardhome-proc-settings.sh:400]: @coderabbitai explain this issue as a prompt for an AI Agent

File: AdGuardHome.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply AdGuardHome.sh]: @coderabbitai explain this issue as a prompt for an AI Agent

**Comment by jumpsmm7** [comment]
@codex implement fixes using this prompt:

```text
Fix a backward-compatibility defect in `AdGuardHome.sh`, in `dnsmasq_ipset_state_recover_pending()`.

Problem:
The current code deletes any `.AdGuardHome.dnsmasq-ipset.*` snapshot that contains `restore.pending` but contains neither `config.pending` nor `config.restored`.

That condition is unsafe for upgrades. Installer versions that predate `config.pending` and `config.restored` create valid rollback snapshots with only:
- `restore.pending`
- `ipset` or `ipset.absent`
- `yaml` or `yaml.absent`

If an older version was interrupted during an IPSET transaction, the snapshot must restore IPSET and YAML state on the next recovery. The new condition deletes this legacy snapshot instead. This leaves the partially updated IPSET and YAML state active and permanently removes the rollback data.

Required change:
1. Preserve the existing recovery behavior for unversioned `restore.pending` snapshots that do not contain `config.pending` or `config.restored`.
2. Add an explicit, durable marker for snapshots that are safe to clean up after dnsmasq-backup creation fails.
3. During recovery, delete only snapshots with that explicit cleanup-only marker.
4. Continue to reject unsafe paths, symlinks, malformed marker combinations, and failed cleanup operations.
5. Do not restore IPSET or YAML data from a snapshot that is explicitly marked cleanup-only.
6. Keep the change compatible with POSIX `/bin/sh` and BusyBox `ash`.

Tests:
- Add a regression test for a legacy snapshot with `restore.pending`, IPSET state, and YAML state, but no `config.pending` or `config.restored`. Recovery must restore the saved IPSET and YAML state.
- Add a test for the new explicit cleanup-only marker. Recovery must remove that snapshot without restoring IPSET or YAML state.
- Add a test that forces cleanup failure for a cleanup-only snapshot. Recovery must fail and retain the snapshot.
- Run the focused dnsmasq/IPSET tests and `sh -n AdGuardHome.sh`.
```

File: tests/dnsmasq-lan-mode.sh [hunk 3]
```diff
@@ -321,8 +331,13 @@ mv() {
 
 # rm injects snapshot-finalization cleanup failures and delegates all other removals to the system command.
 rm() {
+	if [ "${RESTORE_MARKER_REMOVE_FAIL:-0}" = "1" ] && [ "${1:-}" = -f ] &&
+		[ "${2:-}" = "${RESTORE_MARKER_REMOVE_FAIL_SNAPSHOT:-}/restore.pending" ]; then
+		return 1
+	fi
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/dnsmasq-lan-mode.sh:337]: @coderabbitai explain this issue as a prompt for an AI Agent

File: tests/dnsmasq-lan-mode.sh [hunk 4]
```diff
@@ -304,6 +304,16 @@ mv() {
...
+	if [ "${ASSOCIATION_TRANSITION_FAIL:-0}" = "1" ] &&
+		[ "${1:-}" = "${IPSET_SNAPSHOT_DIR:-}/cleanup.only" ] &&
+		[ "${2:-}" = "${IPSET_SNAPSHOT_DIR:-}/restore.pending" ]; then
+		return 1
+	fi
+	if [ "${CLEANUP_ONLY_CREATE_FAIL:-0}" = "1" ] && [ "${2:-}" = "${IPSET_SNAPSHOT_DIR:-}/cleanup.only" ]; then
+		case "${1:-}" in
+			"${IPSET_SNAPSHOT_DIR}/cleanup.only."*) return 1 ;;
+		esac
+	fi
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/dnsmasq-lan-mode.sh:316]: @coderabbitai explain this issue as a prompt for an AI Agent

**Comment by jumpsmm7** [comment]
@codex implement these changes:
Verify and fix the missing failure-injection coverage in `tests/dnsmasq-lan-mode.sh`.

Context:
- The `mv()` test stub contains a failure branch controlled by
  `CLEANUP_ONLY_CREATE_FAIL`.
- That branch returns failure when production code attempts to publish a staged
  `cleanup.only` marker into `IPSET_SNAPSHOT_DIR`.
- The current test scenarios do not assign `CLEANUP_ONLY_CREATE_FAIL='1'`.
- As a result, the test suite does not exercise the production path where
  creation of the durable `cleanup.only` marker fails.

Required work:
1. Inspect `AdGuardHome.sh` and `tests/dnsmasq-lan-mode.sh`.
2. Identify the production path that creates `cleanup.only`.
3. Add or extend one focused test scenario that sets
   `CLEANUP_ONLY_CREATE_FAIL='1'` before that path runs.
4. Verify that the operation returns failure.
5. Verify that the test does not report transaction success.
6. Verify that IPSET, YAML, and dnsmasq state remain safe for the specific
   transaction stage that fails.
7. Verify that no unsafe snapshot marker combination remains.
8. Reset `CLEANUP_ONLY_CREATE_FAIL` in `reset_case()` to prevent state leakage
   into later test scenarios.
9. Keep the change limited to regression coverage unless production behavior is
   incorrect.
10. Run `sh -n tests/dnsmasq-lan-mode.sh` after the change.

Do not assume the review finding is correct. First confirm that no existing test
assigns `CLEANUP_ONLY_CREATE_FAIL='1'` and reaches the matching `mv()` branch.


Review `tests/dnsmasq-lan-mode.sh`.

The test fixture uses failure-injection variables to make the mocked `rm()` and `mv()` functions fail during specific snapshot transitions.

Inspect `reset_case()`. It resets several failure-injection variables, but it does not reset these variables:

- `RESTORE_MARKER_REMOVE_FAIL`
- `RESTORE_MARKER_REMOVE_FAIL_SNAPSHOT`
- `ASSOCIATION_TRANSITION_FAIL`
- `CLEANUP_ONLY_CREATE_FAIL`

This can create test-order dependency. If one test enables one of these variables and a later test calls `reset_case()` without explicitly clearing it, the later test can receive an unrelated injected `rm` or `mv` failure. The reported failure location will then be misleading.

Update `reset_case()` to restore each listed variable to its default value. Set boolean flags to `0` and the snapshot-path variable to an empty string.

Keep the change limited to fixture state reset logic. Validate that the test file remains valid POSIX `/bin/sh` syntax.

**Comment by jumpsmm7** [comment]
@coderabbitai implement these changes:
Verify and fix the missing failure-injection coverage in `tests/dnsmasq-lan-mode.sh`.

Context:
- The `mv()` test stub contains a failure branch controlled by
  `CLEANUP_ONLY_CREATE_FAIL`.
- That branch returns failure when production code attempts to publish a staged
  `cleanup.only` marker into `IPSET_SNAPSHOT_DIR`.
- The current test scenarios do not assign `CLEANUP_ONLY_CREATE_FAIL='1'`.
- As a result, the test suite does not exercise the production path where
  creation of the durable `cleanup.only` marker fails.

Required work:
1. Inspect `AdGuardHome.sh` and `tests/dnsmasq-lan-mode.sh`.
2. Identify the production path that creates `cleanup.only`.
3. Add or extend one focused test scenario that sets
   `CLEANUP_ONLY_CREATE_FAIL='1'` before that path runs.
4. Verify that the operation returns failure.
5. Verify that the test does not report transaction success.
6. Verify that IPSET, YAML, and dnsmasq state remain safe for the specific
   transaction stage that fails.
7. Verify that no unsafe snapshot marker combination remains.
8. Reset `CLEANUP_ONLY_CREATE_FAIL` in `reset_case()` to prevent state leakage
   into later test scenarios.
9. Keep the change limited to regression coverage unless production behavior is
   incorrect.
10. Run `sh -n tests/dnsmasq-lan-mode.sh` after the change.

Do not assume the review finding is correct. First confirm that no existing test
assigns `CLEANUP_ONLY_CREATE_FAIL='1'` and reaches the matching `mv()` branch.


Review `tests/dnsmasq-lan-mode.sh`.

The test fixture uses failure-injection variables to make the mocked `rm()` and `mv()` functions fail during specific snapshot transitions.

Inspect `reset_case()`. It resets several failure-injection variables, but it does not reset these variables:

- `RESTORE_MARKER_REMOVE_FAIL`
- `RESTORE_MARKER_REMOVE_FAIL_SNAPSHOT`
- `ASSOCIATION_TRANSITION_FAIL`
- `CLEANUP_ONLY_CREATE_FAIL`

This can create test-order dependency. If one test enables one of these variables and a later test calls `reset_case()` without explicitly clearing it, the later test can receive an unrelated injected `rm` or `mv` failure. The reported failure location will then be misleading.

Update `reset_case()` to restore each listed variable to its default value. Set boolean flags to `0` and the snapshot-path variable to an empty string.

Keep the change limited to fixture state reset logic. Validate that the test file remains valid POSIX `/bin/sh` syntax.
